cpp/src/ffi/reader_writer_jni.cpp

```diff
@@ -0,0 +1,512 @@
+    env->ThrowNew(exc_class, "Failed to flush writer");
+  }
+}
+
+JNIEXPORT jstring JNICALL
+Java_io_milvus_storage_MilvusStorageWriter_writerClose(JNIEnv* env, jobject obj, jlong writer_handle) {
+  try {
+    WriterHandle handle = static_cast<WriterHandle>(writer_handle);
+
+    char* manifest = nullptr;
+    size_t manifest_size = 0;
+    FFIResult result = writer_close(handle, &manifest, &manifest_size);
+
+    if (!IsSuccess(&result)) {
+      ThrowJavaExceptionFromFFIResult(env, &result);
+      FreeFFIResult(&result);
+      return nullptr;
+    }
+
+    jstring java_manifest = env->NewStringUTF(manifest);
+    free_manifest(manifest);
+
+    return java_manifest;
+  } catch (...) {
+    jclass exc_class = env->FindClass("java/lang/RuntimeException");
+    env->ThrowNew(exc_class, "Failed to close writer");
+    return nullptr;
+  }
+}
+
+JNIEXPORT void JNICALL
+Java_io_milvus_storage_MilvusStorageWriter_writerDestroy(JNIEnv* env, jobject obj, jlong writer_handle) {
+  try {
+    WriterHandle handle = static_cast<WriterHandle>(writer_handle);
+    writer_destroy(handle);
+  } catch (...) {
+    jclass exc_class = env->FindClass("java/lang/RuntimeException");
+    env->ThrowNew(exc_class, "Failed to destroy writer");
+  }
+}
+
+// ==================== JNI Reader Implementation ====================
+
+JNIEXPORT jlong JNICALL
+Java_io_milvus_storage_MilvusStorageReader_readerNew(JNIEnv* env, jobject obj, jstring manifest, jlong schema_ptr, jobjectArray needed_columns, jlong properties_ptr) {
+  try {
+    const char* manifest_cstr = env->GetStringUTFChars(manifest, nullptr);
+    ArrowSchema* schema = reinterpret_cast<ArrowSchema*>(schema_ptr);
+    Properties* properties = reinterpret_cast<Properties*>(properties_ptr);
+
+    size_t num_columns = 0;
+    const char** columns = ConvertFromJavaStringArray(env, needed_columns, &num_columns);
+
+    ReaderHandle reader_handle;
+    FFIResult result = reader_new(const_cast<char*>(manifest_cstr), schema, columns, num_columns, properties, &reader_handle);
+
+    env->ReleaseStringUTFChars(manifest, manifest_cstr);
+    FreeStringArray(env, columns, num_columns);
+
+    if (!IsSuccess(&result)) {
+      ThrowJavaExceptionFromFFIResult(env, &result);
+      FreeFFIResult(&result);
+      return 0;
+    }
+
+    return static_cast<jlong>(reader_handle);
+  } catch (...) {
+    jclass exc_class = env->FindClass("java/lang/RuntimeException");
+    env->ThrowNew(exc_class, "Failed to create reader");
+    return 0;
+  }
+}
+
+JNIEXPORT jlong JNICALL
+Java_io_milvus_storage_MilvusStorageReader_getRecordBatchReader(JNIEnv* env, jobject obj, jlong reader_handle, jstring predicate, jlong batch_size, jlong buffer_size) {
+  try {
+    ReaderHandle handle = static_cast<ReaderHandle>(reader_handle);
+    const char* predicate_cstr = predicate ? env->GetStringUTFChars(predicate, nullptr) : nullptr;
+
+    ArrowArrayStream* stream = static_cast<ArrowArrayStream*>(malloc(sizeof(ArrowArrayStream)));
```

Comment:
in JNI, if we do malloc in this layer, then you should also provider the `free` function.

BTW, in arrow ABI, all struct in ABI need call the `release` function before free itself.
```
ArrowArrayStream* stream = ...;

  if (stream && stream->release)
  {
     stream->release(stream);
  }
  
```